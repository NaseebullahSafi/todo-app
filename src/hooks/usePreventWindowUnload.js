import { useEffect } from "react";

/**
 * custom hook to run before page unload
 * credit: https://angelos.dev/2019/05/custom-react-hook-to-prevent-window-unload/
 * @param {*} deps - dependencies injected to useEffect
 * @param {*} handleBeforeUnload - func. to run before page unloads
 */

export const usePreventWindowUnload = (deps, handleBeforeUnload) => {
  useEffect(() => {
    // the first element of deps is to check if browser is allowed to unload
    if (!deps[0]) return;
    window.addEventListener("beforeunload", handleBeforeUnload);
    return () => window.removeEventListener("beforeunload", handleBeforeUnload);
  }, deps);
};
