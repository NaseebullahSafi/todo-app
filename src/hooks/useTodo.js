import { useState } from "react";
import { useDispatch, useSelector } from "react-redux";

import { deleteTodo, editTodo } from "src/Store/todos/actions";
import {
  addRemoveToStack,
  addEditToStack,
} from "src/Store/userActions/actions";

/**
 * Will format the date and return a new string with month & date
 * @param {String} date
 */
const formatDate = (date) => {
  if (date) {
    const [, month, dateNum] = date.split(" ");
    return `${month} ${dateNum}`;
  }
};

/**
 * Custom hook which abstracts many operations away from the component
 * @param {*} id - the id of the todo
 * @param {*} name - the name of the todo
 * @param {*} description - the description of the todo
 * @param {*} date - the date it was created/modified
 */
export const useTodo = (id, name, description, date) => {
  // local state to determine if its editable
  const [editable, setEditable] = useState(false);

  // local state of todo for tracking changes from user input
  const [todo, updateTodo] = useState({
    id,
    name,
    description,
    date,
  });

  // retrieve userActions stack from redux
  const { userActions } = useSelector(({ userActions }) => ({ userActions }));

  // needed to dispatch actions to redux
  const dispatch = useDispatch();

  const toggleEditable = (v) => setEditable(v);

  /**
   * onChange func. which will update our local state when user is typing
   * @param {*} e
   */
  const onChange = (e) =>
    updateTodo({
      ...todo,
      [e.target.name]: e.target.value,
    });

  /**
   * func. which will submit the form, thereby updating redux and register
   * the action into our stack
   * @param {*} e
   */
  const onSubmit = (e) => {
    e.preventDefault();

    const data = {
      ...todo,
      date: new Date().toString(),
    };

    // update our redux with the new values for respective todo
    dispatch(
      editTodo({
        id,
        data,
      })
    );

    // if user is recording, then also register the action
    if (userActions.isRecording) {
      dispatch(
        addEditToStack({
          id,
          data,
          action: "EDIT",
        })
      );
    }

    // toggle the form back to readable
    setEditable(false);
  };

  /**
   * func. which will delete the todo from redux
   * @param {*} e
   */
  const onDelete = (e) => {
    e.preventDefault();

    dispatch(deleteTodo(id));

    if (userActions.isRecording) {
      dispatch(
        addRemoveToStack({
          id,
          action: "DELETE",
        })
      );
    }
  };

  return {
    editable,
    todo,
    toggleEditable,
    formatDate,
    onChange,
    onDelete,
    onSubmit,
  };
};
