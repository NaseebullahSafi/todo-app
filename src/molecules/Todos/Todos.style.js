import styled from "styled-components";

export const StyledList = styled.div``;

export const StyledPlay = styled.button`
  border-radius: 100%;
  padding: 1em;
  width: 70px;
  height: 70px;
  display: flex;
  align-items: center;
  justify-content: center;
  box-shadow: 0 4px 6px rgba(50, 50, 93, 0.11), 0 1px 3px rgba(0, 0, 0, 0.08);
  transition: all 0.15s ease-in;
  cursor: pointer;
  border-radius: 0.4em;
  background-color: white;
  width: 60px;
  height: 60px;
  margin: 0 0 1em 1em;

  svg {
    color: #525f7f;
    transition: all 0.15s ease-in;
    fill: transparent;
  }

  &:hover {
    svg {
      fill: #f5365c;
      stroke: #f5365c;
    }
  }
`;
