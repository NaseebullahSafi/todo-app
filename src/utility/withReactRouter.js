import React from "react";
import { MemoryRouter } from "react-router-dom";
import { create } from "react-test-renderer";

export const withReactRouter = (component) =>
  create(<MemoryRouter>{component}</MemoryRouter>);
