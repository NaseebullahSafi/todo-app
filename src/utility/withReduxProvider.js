import React from "react";
import { Provider } from "react-redux";

import store from "../Store";

export const ReduxProvider = ({ children }) => (
  <Provider store={store}>{children}</Provider>
);
