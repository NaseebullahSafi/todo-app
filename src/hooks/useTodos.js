import { useState, useCallback, useEffect } from "react";

import { useDispatch, useSelector } from "react-redux";

// local components
import { usePreventWindowUnload } from "src/hooks/usePreventWindowUnload";
import {
  deleteAllTodos,
  fetchAllTodos,
  playbackRecord,
} from "src/Store/todos/actions";

/**
 * Custom hook which abstracts the side-effects from the component.
 */
export const useTodos = () => {
  // keep track of the play state
  const [play, setPlay] = useState(false);

  // get the two reducers from redux
  const { todos, userActions } = useSelector(({ todos, userActions }) => ({
    todos,
    userActions,
  }));

  // needed to dispatch actions in redux
  const dispatch = useDispatch();

  // func. to toggle the state of play
  const togglePlay = useCallback(() => setPlay((p) => !p), []);

  // func. to dispatch a deleteAllTodos action
  const onDelete = () => dispatch(deleteAllTodos(true));

  // which will retrieve latest todos from localStorage on component mount only
  useEffect(() => {
    dispatch(fetchAllTodos());
  }, []);

  // clean all todos if play btn was pressed
  useEffect(() => {
    if (play) {
      // delete call todos, without deleting it from localstorage
      dispatch(deleteAllTodos(false));
    }
  }, [play]);

  // On play, remove from stack, and add to todos state, with 1s delay
  useEffect(() => {
    if (play) {
      // start replaying user actions. Once done, toggle the play state
      dispatch(playbackRecord(userActions, (cb) => togglePlay(cb)));
    }
  }, [play, userActions]);

  /**
   * The first element of array is to tell the hook the browser
   * is allowed to unload. Second element is to update when todos is
   * updated.
   */
  usePreventWindowUnload([true, todos], (e) => {
    localStorage.setItem("todos", JSON.stringify({ ...todos }));
  });

  return { play, togglePlay, todos, userActions, onDelete };
};
