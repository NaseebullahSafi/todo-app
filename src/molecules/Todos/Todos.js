import React from "react";

import Todo from "src/molecules/Todo";
import { StyledList, StyledBtn, StyledRow } from "./Todos.style";
import { useTodos } from "src/hooks/useTodos";

/**
 * Todos component which will retrieve todos from localStorage,
 * and render them.
 */
const Todos = () => {
  // using custom hook to abstract the side effects of updates from redux
  const { togglePlay, todos, onDelete } = useTodos();

  return (
    <StyledList>
      <StyledRow>
        <StyledBtn
          disable={Object.keys(todos).length === 0}
          onClick={() => togglePlay(true)}
        >
          <svg
            xmlns="http://www.w3.org/2000/svg"
            width="16"
            height="16"
            viewBox="0 0 24 24"
            fill="none"
            stroke="currentColor"
            strokeWidth="2"
            strokeLinecap="round"
            strokeLinejoin="round"
            className="feather feather-play"
          >
            <polygon points="5 3 19 12 5 21 5 3" />
          </svg>
        </StyledBtn>
        <StyledBtn
          disableIconHover
          disable={Object.keys(todos).length === 0}
          onClick={onDelete}
        >
          <svg
            xmlns="http://www.w3.org/2000/svg"
            width="16"
            height="16"
            viewBox="0 0 24 24"
            fill="none"
            stroke="currentColor"
            strokeWidth="2"
            strokeLinecap="round"
            strokeLinejoin="round"
            className="feather feather-trash-2"
          >
            <polyline points="3 6 5 6 21 6" />
            <path d="M19 6v14a2 2 0 0 1-2 2H7a2 2 0 0 1-2-2V6m3 0V4a2 2 0 0 1 2-2h4a2 2 0 0 1 2 2v2" />
            <line x1="10" y1="11" x2="10" y2="17" />
            <line x1="14" y1="11" x2="14" y2="17" />
          </svg>
        </StyledBtn>
      </StyledRow>
      {Object.keys(todos).map((k) => {
        const todo = todos[k];
        return <Todo {...todo} key={k} />;
      })}
    </StyledList>
  );
};

export default Todos;
