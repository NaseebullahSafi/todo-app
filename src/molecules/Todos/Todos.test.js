import React from "react";
import { renderHook, act } from "@testing-library/react-hooks";
import { create } from "react-test-renderer";
import "jest-styled-components";

import reducers from "src/Store/reducer";
import { ReduxProvider } from "src/utility/withReduxProvider";
import { useTodos } from "src/hooks/useTodos";
import Todos from "./Todos";

describe("<Todos />", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = ({ children }) => <ReduxProvider>{children}</ReduxProvider>;
  });

  it("should render Todos", () => {
    const tree = create(
      <ReduxProvider>
        <Todos />
      </ReduxProvider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should use play", () => {
    const { result } = renderHook(() => useTodos(), { wrapper });
    expect(result.current.play).toBe(false);
  });

  it("should update play", () => {
    const { result } = renderHook(() => useTodos(), { wrapper });

    act(() => {
      result.current.togglePlay();
    });

    expect(result.current.play).toBe(true);
  });

  describe("Redux", () => {
    it("should return initial state of redux", () => {
      const state = reducers(undefined, {});
      expect(state).toEqual({ todos: {}, userActions: {} });
    });

    it("should initially not be recording on app mount", () => {
      const { userActions } = reducers(
        { todos: {}, userActions: {} },
        { type: "ACTION_STOP_RECORDING", payload: false }
      );

      expect(userActions).toEqual({ isRecording: false });
    });

    it("should start recording when clicking on play button", () => {
      const { userActions } = reducers(
        { todos: {}, userActions: {} },
        { type: "ACTION_START_RECORDING", payload: true }
      );

      expect(userActions).toEqual({ isRecording: true });
    });

    it("should delete todo when clicking delete", () => {
      const { todos } = reducers(
        {
          todos: {
            "7b2a3db5-af85-42a5-9440-fa4dc6c7e12b": {
              id: "7b2a3db5-af85-42a5-9440-fa4dc6c7e12b",
              name: "1",
              description: "1",
              date: "Sat Apr 25 2020 16:26:23 GMT+0100 (British Summer Time)",
            },
          },
          userActions: {
            isRecording: false,
            actions: [
              { id: "7b2a3db5-af85-42a5-9440-fa4dc6c7e12b", action: "ADD" },
            ],
          },
        },
        { type: "TODO_DELETE", payload: "7b2a3db5-af85-42a5-9440-fa4dc6c7e12b" }
      );
      expect(todos).toEqual({});
    });

    it("should not record the delete action if app is not in record state", () => {
      const { userActions } = reducers(
        {
          todos: {
            "7b2a3db5-af85-42a5-9440-fa4dc6c7e12b": {
              id: "7b2a3db5-af85-42a5-9440-fa4dc6c7e12b",
              name: "1",
              description: "1",
              date: "Sat Apr 25 2020 16:26:23 GMT+0100 (British Summer Time)",
            },
          },
          userActions: {
            isRecording: false,
            actions: [
              { id: "7b2a3db5-af85-42a5-9440-fa4dc6c7e12b", action: "ADD" },
            ],
          },
        },
        { type: "TODO_DELETE", payload: "7b2a3db5-af85-42a5-9440-fa4dc6c7e12b" }
      );
      expect(userActions).toEqual({
        isRecording: false,
        actions: [
          { id: "7b2a3db5-af85-42a5-9440-fa4dc6c7e12b", action: "ADD" },
        ],
      });
    });

    it("should record the delete action when app is in record state", () => {
      const { userActions } = reducers(
        { todos: {}, userActions: { isRecording: true } },
        {
          type: "ACTION_ADD_DELETE",
          payload: {
            id: "7b2a3db5-af85-42a5-9440-fa4dc6c7e12b",
            action: "DELETE",
          },
        }
      );
      expect(userActions).toEqual({
        isRecording: true,
        actions: [
          { id: "7b2a3db5-af85-42a5-9440-fa4dc6c7e12b", action: "DELETE" },
        ],
      });
    });
  });
});
