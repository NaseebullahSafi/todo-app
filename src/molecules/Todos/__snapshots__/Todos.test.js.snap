// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Todos /> should render Todos 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c1 {
  border-radius: 100%;
  padding: 1em;
  width: 70px;
  height: 70px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  box-shadow: 0 4px 6px rgba(50,50,93,0.11),0 1px 3px rgba(0,0,0,0.08);
  -webkit-transition: all 0.15s ease-in;
  transition: all 0.15s ease-in;
  cursor: pointer;
  border-radius: 0.4em;
  background-color: white;
  width: 60px;
  height: 60px;
  margin: 1em;
  opacity: 0.5;
  pointer-events: none;
  -webkit-transform: scale(0.9);
  -ms-transform: scale(0.9);
  transform: scale(0.9);
  cursor: none;
}

.c1 svg {
  color: #525f7f;
  -webkit-transition: all 0.15s ease-in;
  transition: all 0.15s ease-in;
  fill: transparent;
}

.c1:hover svg {
  fill: #f5365c;
  stroke: #f5365c;
}

.c2 {
  border-radius: 100%;
  padding: 1em;
  width: 70px;
  height: 70px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  box-shadow: 0 4px 6px rgba(50,50,93,0.11),0 1px 3px rgba(0,0,0,0.08);
  -webkit-transition: all 0.15s ease-in;
  transition: all 0.15s ease-in;
  cursor: pointer;
  border-radius: 0.4em;
  background-color: white;
  width: 60px;
  height: 60px;
  margin: 1em;
  opacity: 0.5;
  pointer-events: none;
  -webkit-transform: scale(0.9);
  -ms-transform: scale(0.9);
  transform: scale(0.9);
  cursor: none;
}

.c2 svg {
  color: #525f7f;
  -webkit-transition: all 0.15s ease-in;
  transition: all 0.15s ease-in;
  fill: transparent;
}

<div
  className=""
>
  <div
    className="c0"
  >
    <button
      className="c1"
      onClick={[Function]}
    >
      <svg
        className="feather feather-play"
        fill="none"
        height="16"
        stroke="currentColor"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth="2"
        viewBox="0 0 24 24"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polygon
          points="5 3 19 12 5 21 5 3"
        />
      </svg>
    </button>
    <button
      className="c2"
      onClick={[Function]}
    >
      <svg
        className="feather feather-trash-2"
        fill="none"
        height="16"
        stroke="currentColor"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth="2"
        viewBox="0 0 24 24"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polyline
          points="3 6 5 6 21 6"
        />
        <path
          d="M19 6v14a2 2 0 0 1-2 2H7a2 2 0 0 1-2-2V6m3 0V4a2 2 0 0 1 2-2h4a2 2 0 0 1 2 2v2"
        />
        <line
          x1="10"
          x2="10"
          y1="11"
          y2="17"
        />
        <line
          x1="14"
          x2="14"
          y1="11"
          y2="17"
        />
      </svg>
    </button>
  </div>
</div>
`;
